pkg/operator/csr/csr_approver.go

```diff
@@ -0,0 +1,180 @@
+package webhookcerts
+
+import (
+	"context"
+	"crypto/x509"
+	"encoding/pem"
+	"fmt"
+
+	certapiv1 "k8s.io/api/certificates/v1"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/apiserver/pkg/authentication/serviceaccount"
+	certv1informers "k8s.io/client-go/informers/certificates/v1"
+	certv1client "k8s.io/client-go/kubernetes/typed/certificates/v1"
+	certv1listers "k8s.io/client-go/listers/certificates/v1"
+
+	"github.com/openshift/library-go/pkg/controller/factory"
+	"github.com/openshift/library-go/pkg/operator/events"
+	"github.com/openshift/library-go/pkg/operator/v1helpers"
+)
+
+type CSRApprovalDecision string
+
+const (
+	CSRApproved  CSRApprovalDecision = "Approved"
+	CSRDenied    CSRApprovalDecision = "Denied"
+	CSRNoOpinion CSRApprovalDecision = "NoOpinion"
+)
+
+type CSRApproveFunc func(csrObj *certapiv1.CertificateSigningRequest, x509CSR *x509.CertificateRequest) (approvalStatus CSRApprovalDecision, denyReason string, err error)
+
+type csrApproverController struct {
+	csrClient certv1client.CertificateSigningRequestInterface
+	csrLister certv1listers.CertificateSigningRequestLister
+
+	csrName     string
+	csrApprover CSRApproveFunc
+}
+
+// NewCSRApproverController returns a controller that is observing the CSR API
+// for a CSR of a given name. If such a CSR exists, it runs the `csrApprover` against
+// it and either denies, approves or leaves the CSR.
+func NewCSRApproverController(
+	operatorClient v1helpers.OperatorClient,
+	csrClient certv1client.CertificateSigningRequestInterface,
+	csrInformers certv1informers.CertificateSigningRequestInformer,
+	csrName string,
+	csrApprover CSRApproveFunc,
+	eventsRecorder events.Recorder,
+) factory.Controller {
+	c := &csrApproverController{
+		csrClient:   csrClient,
+		csrLister:   csrInformers.Lister(),
+		csrName:     csrName,
+		csrApprover: csrApprover,
```

Comment:
minor: `IsApproved` or `csrApproverFn` or some such please.  Not a noun name.

Reply:
I'm now using an interface instead of a function as suggested below so I kept the name